<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Bandage by rrwick</title>

    <link rel="icon" type="image/png" href="favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="favicon-16x16.png" sizes="16x16" />

  </head>

  <body>
    <header>
	<div class="inner">
      <p style="float: left;">
	    <img src="images/logo.png" class="logo">
      </p>
      <div>
        <h1>Bandage</h1>
        <h2>a <u>B</u>ioinformatics <u>A</u>pplication for <u>N</u>avigating <i><u>D</u>e novo</i> <u>A</u>ssembly <u>G</u>raphs <u>E</u>asily
      </div>
    </div>
      
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
		
<p>Bandage is a program for visualising <em>de novo</em> assembly graphs. By displaying connections which are not present in the contigs file, Bandage opens up new possibilities for analysing <em>de novo</em> assemblies.</p>

<div style="text-align: center;">
<img src="images/wiki/bandage_gui.png" style="max-width:80%;"><br>
</div>



<h2>
<a id="features" class="anchor" href="#features" aria-hidden="true"><span class="octicon octicon-link"></span></a>Features</h2>

<ul class="list-a">
  <li>Load multiple assembly graph formats: LastGraph (Velvet), FASTG (SPAdes), Trinity.fasta and GFA.</li>
  <li>Position nodes automatically with an efficient graph layout algorithm.</li>
  <li>Zoom, pan and rotate the view using either mouse or keyboard controls.</li>
  <li>Reposition and reshape nodes by clicking and dragging with the mouse.</li>
  <li>Configure graph scope: view the entire assembly graph or only a region of interest.</li>
  <li>Copy node sequences to the clipboard or save them to file.</li>
  <li>Colour nodes using built-in colour schemes or user-defined colours.</li>
  <li>Label nodes using node number, length, coverage or user-defined labels.</li>
  <li>Find nodes quickly in a large graph using node numbers.</li>
  <li>Specify the thickness of nodes and allow thickness to reflect the node's read depth.</li>
  <li>Define the relationship between the length of a node and the length of its sequence.</li>
  <li>Draw graph in single node style: each node and its reverse complements appear as a single object.</li>
  <li>Draw graph in double node style: nodes and their reverse complements appear as separate objects with arrow heads to indicate direction.</li>
  <li>Highlight and label specific sequences with integrated BLAST search.</li>
  <li>Automatically identify nodes contiguous with a node of interest.</li>
  <li>Call Bandage from the command line to specify settings, load graphs or generate images.</li>
</ul>


<h2>
<a id="uses" class="anchor" href="#uses" aria-hidden="true"><span class="octicon octicon-link"></span></a>Uses</h2>

<ul class="list-a">
  <li>Assess and compare the quality of assemblies quickly and visually.</li>
  <li>Identify problematic regions of an assembly.</li>
  <li>Resolve ambiguities in the graph to improve or complete <i>de novo</i> assemblies.</li>
  <li>Determine which other nodes have sequences that are contiguous with a node of interest.</li>
  <li>Annotate graph images to illustrate assembly features.</li>
  <li>Extract candidate sequences that extend through multiple nodes.</li>
</ul>


<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Bandage works equally well on OS X, Linux and Windows. Users are encouraged to download 64-bit binary executables using the links on the top-right of this page. No installation is necessary â€“ just unzip and run. Windows and Mac binaries come packaged with all necessary libraries. The Linux binary has a couple of dependencies, specified in the zip file.  More pre-built binaries are available on the <a href="https://github.com/rrwick/Bandage/releases/" target="_blank">GitHub 'Releases' page</a>.</p>

<p>When first opening Bandage on a Mac, you may receive a warning stating that Bandage 'can't be opened because it is from an unidentified developer.'  Right click on the file and select 'Open' to override this warning.</p>

<p>Alternatively, you can clone the source code from GitHub and build Bandage yourself. Instructions are on <a href="https://github.com/rrwick/Bandage/" target="_blank"> the GitHub page</a>.</p>


<h2>
<a id="sampledata" class="anchor" href="#sampledata" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sample data</h2>

<p>Here are a couple of sample graph files you can download and view in Bandage:</p>

<ul class="list-a">
<li><a href="samples/E_coli_LastGraph.zip" target="_blank">An <i>E. coli</i> assembly from Velvet</a>. The reads and more information are available <a href="http://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=272917" target="_blank">here</a>.</li>
<li><a href="samples/Trinity.fasta.zip" target="_blank">A small mouse transcriptome assembly from Trinity</a>. This data comes with Trinity in the 'sample_data/test_Trinity_Assembly/' folder.</li>
</ul>

<h2>
<a id="documentation" class="anchor" href="#documentation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Documentation</h2>

<p>Bandage's documentation is on the <a href="https://github.com/rrwick/Bandage/wiki" target="_blank"> Bandage GitHub wiki</a>.</p>


<h2>
<a id="cite" class="anchor" href="#cite" aria-hidden="true"><span class="octicon octicon-link"></span></a>Cite</h2>

<p>If you use Bandage in your research, please cite the following publication:</p>
<p><a href="http://bioinformatics.oxfordjournals.org/content/31/20/3350">Wick R.R., Schultz M.B., Zobel J. & Holt K.E. (2015). Bandage: interactive visualisation of <em>de novo</em> genome assemblies. <em>Bioinformatics</em>, 31(20), 3350-3352.</a> </p>


        </section>

        <aside id="sidebar">
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.7.1/Bandage_Mac_v0.7.1.zip" class="button_mac">
            <small>Download</small>
            Mac
          </a>
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.7.1/Bandage_Ubuntu_14.04_dynamic_v0.7.1.zip" class="button_linux">
            <small>Download</small>
            Linux
          </a>
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.7.1/Bandage_Windows_v0.7.1.zip" class="button_windows">
            <small>Download</small>
            Windows
          </a>
          <a href="https://github.com/rrwick/Bandage" class="button_github">
            <small>View project on</small>
            GitHub
          </a>
          <a href="https://github.com/rrwick/Bandage/wiki" class="button_wiki">
            <small>Documentation</small>
            Wiki
          </a>

          <br><br>

          <p class="repo-owner"><a href="https://github.com/rrwick/Bandage"></a>Bandage is made by <a href="https://github.com/rrwick">Ryan Wick</a> in <a href="http://holtlab.net/">Kathryn Holt's research group</a>.</p>

          <p align="center">
            <a href="http://iawards.com.au/index.php/winners/2015-winners/2015-national-winners" target="_blank"> <img src="images/iawards.png" alt="iAwards 2015 national winner" width="100" height="100"></a>
          </p>

        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-59453959-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
