<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Bandage by rrwick</title>

    <link rel="icon" type="image/png" href="favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="favicon-16x16.png" sizes="16x16" />

  </head>

  <body>
    <header>
	<div class="inner">
      <p style="float: left;">
	    <img src="images/logo.png" class="logo">
      </p>
      <div>
        <h1>Bandage</h1>
        <h2>a <u>B</u>ioinformatics <u>A</u>pplication for <u>N</u>avigating <i><u>D</u>e novo</i> <u>A</u>ssembly <u>G</u>raphs <u>E</u>asily
      </div>
    </div>
      
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
		
		
	
		  
<p><em>De novo</em> assembly graphs contain assembled contigs (nodes) but also the connections between those contigs (edges), which are not easily accessible to users.  Bandage is a program for visualising assembly graphs using graph layout algorithms.  By displaying connections between contigs, Bandage opens up new possibilities for analysing <em>de novo</em> assemblies that are not possible by looking at contigs alone.</p>





<h2>
<a id="features" class="anchor" href="#features" aria-hidden="true"><span class="octicon octicon-link"></span></a>Features</h2>

<ul class="list-a">
  <li>A layout algorithm is used to automatically position graph nodes.</li>
  <li>Manually reposition and reshape nodes.</li>
  <li>Zoom, pan and rotate the view using either mouse or keyboard controls.</li>
  <li>View the entire assembly graph or only a region of interest.</li>
  <li>Copy node sequences to the clipboard or save them to file.</li>
  <li>Nodes can be coloured using built-in colour schemes or user-defined colours.</li>
  <li>Nodes can be labelled using node number, length, coverage or a user-defined label.</li>
  <li>Find nodes quickly in a large graph using node numbers.</li>
  <li>Specify the thickness of nodes and allow thickness to reflect the node's coverage.</li>
  <li>Define the relationship between the length of a node and the length of its sequence.</li>
  <li>Two possible styles for handling reverse complement nodes:
    <ul class="list-a">
      <li>Single: nodes and their reverse complements are drawn as one object with no direction.</li>
      <li>Double: nodes are drawn using arrow heads to indicate direction and reverse complement nodes are drawn separately with arrow heads pointing the opposite direction.</li>
    </ul>
  </li>
  <li>Integrated BLAST search allows for highlighting specific sequences in an assembly graph.</li>
</ul>


<h2>
<a id="uses" class="anchor" href="#uses" aria-hidden="true"><span class="octicon octicon-link"></span></a>Uses</h2>

<ul class="list-a">
  <li>Assess and compare the quality of assemblies quickly and visually.</li>
  <li>Identify problematic regions of an assembly.</li>
  <li>Resolve ambiguities in the graph to improve or complete <i>de novo</i> assemblies.</li>
  <li>Determine which other nodes have sequences that are contiguous with a node of interest.</li>
  <li>Annotate graph images to illustrate assembly features.</li>
  <li>Extract candidate sequences that extend through multiple nodes.</li>
</ul>


<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Bandage works equally well on Windows, OS X and Linux.  Users are encouraged to download 64-bit binary executables using the links on this page or from the <a href="https://github.com/rrwick/Bandage/releases/" target="_blank">GitHub 'Releases' page</a>.  No installation is necessary â€“ just unzip and run.  Windows and Mac binaries come packaged with all necessary libraries.  The Linux binary has a couple of dependencies, specified in the zip file.</p>

<p>Alternatively, you can clone the source code from GitHub and build Bandage yourself.  Instructions are on <a href="https://github.com/rrwick/Bandage/" target="_blank"> the GitHub page</a>.</p>




<h2>
<a id="example" class="anchor" href="#example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example</h2>

<p>For an easy to understand case, let's use a hypothetical bacterial genome that contains a single repeated element (though for the moment, this genome is unknown to the user).  We then simulate 100 bp reads from this sequence and assemble using Velvet.</p>

<div style="text-align: center;">
<div class="diagram" style="max-width:100%; margin-bottom: 20px;">
<img src="images/simple1.png" class="diagram"><br>
</div>
</div>

<p>Because the repeated element is longer than the sequencing reads, the assembler was not able to reproduce the original genome as a single contig.  Rather, three contigs are produced: one for the repeated element (even though it occurs twice) and one for each sequence between the repeated elements.</p>

<p>Given only the contigs, the relationship between these sequences is not clear.  However, the assembly graph contains additional information, and when visualised in Bandage, it is made apparent.</p>

<div style="text-align: center;">

<div class="diagram" style="max-width:25%; margin-right: 5%; margin-left: %; margin-bottom: 20px;">
<img src="images/simple2.png" class="diagram"><br>
<small-green>The assembly graph, as it appears in Bandage</small-green>
</div>

<div class="diagram" style="max-width:55%; margin-right: 5%; margin-left: 5%; margin-bottom: 20px;">
<img src="images/simple3.png" class="diagram"><br>
<small-green>Possible underlying sequences compatible with the assembly graph</small-green>
</div>

</div>


<p style="margin: 0">There are two possible underlying sequences compatible with this graph:</p>
<ul class="list-a" style="margin: 5px">
  <li>Two separate circular sequences that share a region in common</li>
  <li>A single larger circular sequence with an element that occurs twice</li>
</ul>
<p>Additional information, such as prior knowledge on the approximate size of the genome, can help the user to rule out the first alternative.  In this way, Bandage has assisted in turning a fragmented assembly of three contigs into a completed genome of one sequence.</p>



<h2>
<a id="videos" class="anchor" href="#videos" aria-hidden="true"><span class="octicon octicon-link"></span></a>Videos</h2>


<div class="wrapper">
    <div class="h_iframe">
        <!-- a transparent image is preferable -->
        <img class="ratio" src="http://placehold.it/16x9"/>
        <iframe src="https://www.youtube.com/embed/cierloa5hS0?rel=0" frameborder="0" allowfullscreen></iframe>
    </div>
</div>


<h2>
<a id="screenshots" class="anchor" href="#screenshots" aria-hidden="true"><span class="octicon octicon-link"></span></a>Screenshots</h2>

<p>These graphs are the same bacterial isolate assembled in SPAdes using three different k-mer values:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot01.png" target="_blank">
<img src="images/screenshots/screenshot01_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>K-mer of 21.  This value is too small, resulting in short contigs and many connections, giving a dense tangled graph.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot02.png" target="_blank">
<img src="images/screenshots/screenshot02_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>K-mer of 77.  This is a good balance, giving long contigs that are well connected.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot03.png" target="_blank">
<img src="images/screenshots/screenshot03_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>K-mer of 127.  This value is too large, resulting in the graph breaking into many discontinuous pieces.</small-green>
</div>

<p></p>
<p>The following two images illustrate the difference between single and double nodes:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot04.png" target="_blank">
<img src="images/screenshots/screenshot04_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Single nodes, each holding both forward and reverse sequences.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot05.png" target="_blank">
<img src="images/screenshots/screenshot05_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Double nodes, where forward and reverse nodes are drawn separately.</small-green>
</div>

<p></p>
<p>Examples of custom colouring and labelling of nodes in Bandage:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot06.png" target="_blank">
<img src="images/screenshots/screenshot06_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>A bacterial transposon that shares a gene with the cloning vector.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot07.png" target="_blank">
<img src="images/screenshots/screenshot07_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Assembly graph from metagenomic reads that align to particular 16S genes.</small-green>
</div>

<p></p>
<p>A very large graph from a metagenome assembly, displayed in Bandage:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot08.png" target="_blank">
<img src="images/screenshots/screenshot08_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Zoomed out view, showing large-scale structure.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot09.png" target="_blank">
<img src="images/screenshots/screenshot09_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Zoomed in view of smaller subgraphs.</small-green>
</div>

<p></p>
<p>A BLAST search for a 58 kb transposon in a bacterial assembly:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot10.png" target="_blank">
<img src="images/screenshots/screenshot10_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>The entire assembly graph.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot11.png" target="_blank">
<img src="images/screenshots/screenshot11_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Reduced scope to only nodes containing BLAST hits.</small-green>
</div>


<p></p>



        </section>

        <aside id="sidebar">
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.4.2/Bandage_Linux_v0.4.2.zip" class="button_linux">
            <small>Download</small>
            Linux
          </a>
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.4.2/Bandage_Mac_v0.4.2.zip" class="button_mac">
            <small>Download</small>
            Mac
          </a>
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.4.2/Bandage_Windows_v0.4.2.zip" class="button_windows">
            <small>Download</small>
            Windows
          </a>
		  <a href="https://github.com/rrwick/Bandage" class="button_github">
		    <small>View project on</small>
			GitHub
	      </a>

          <p class="repo-owner"><a href="https://github.com/rrwick/Bandage"></a>Bandage is made by <a href="https://github.com/rrwick">Ryan Wick</a>.</p>
        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-59453959-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
