<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/pygment_trac.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>Bandage by rrwick</title>

    <link rel="icon" type="image/png" href="favicon-32x32.png" sizes="32x32" />
    <link rel="icon" type="image/png" href="favicon-16x16.png" sizes="16x16" />

  </head>

  <body>
    <header>
	<div class="inner">
      <p style="float: left;">
	    <img src="images/logo.png" class="logo">
      </p>
      <div>
        <h1>Bandage</h1>
        <h2>a <u>B</u>ioinformatics <u>A</u>pplication for <u>N</u>avigating <i><u>D</u>e novo</i> <u>A</u>ssembly <u>G</u>raphs <u>E</u>asily
      </div>
    </div>
      
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
		
		
	
<p><em>De novo</em> genome assemblers work by building an assembly graph.  Different assemblers represent this graph in different ways, but they all contain sequences (which become assembled contigs) and connections between those sequences.  In the past, assembly graphs have not been easily accessible, and most users only use the assembled contigs, not the graph.  Bandage is a program for visualising assembly graphs. By displaying connections which are not present in the contigs file, Bandage opens up new possibilities for analysing <em>de novo</em> assemblies.</p>



<h2>
<a id="features" class="anchor" href="#features" aria-hidden="true"><span class="octicon octicon-link"></span></a>Features</h2>

<ul class="list-a">
  <li>Load multiple assembly graph formats: LastGraph (Velvet), FASTG (SPAdes), Trinity.fasta and GFA.</li>
  <li>Position nodes automatically with an efficient graph layout algorithm.</li>
  <li>Zoom, pan and rotate the view using either mouse or keyboard controls.</li>
  <li>Reposition and reshape nodes by clicking and dragging with the mouse.</li>
  <li>Configure graph scope: view the entire assembly graph or only a region of interest.</li>
  <li>Copy node sequences to the clipboard or save them to file.</li>
  <li>Colour nodes using built-in colour schemes or user-defined colours.</li>
  <li>Label nodes using node number, length, coverage or user-defined labels.</li>
  <li>Find nodes quickly in a large graph using node numbers.</li>
  <li>Specify the thickness of nodes and allow thickness to reflect the node's coverage.</li>
  <li>Define the relationship between the length of a node and the length of its sequence.</li>
  <li>Draw graph in single node style: each node and its reverse complements appear as a single object.</li>
  <li>Draw graph in double node style: nodes and their reverse complements appear as separate objects with arrow heads to indicate direction.</li>
  <li>Highlight and label specific sequences with integrated BLAST search.</li>
  <li>Automatically identify nodes contiguous with a node of interest.</li>
  <li>Call Bandage from the command line to specify settings, load graphs or generate images.</li>
</ul>


<h2>
<a id="uses" class="anchor" href="#uses" aria-hidden="true"><span class="octicon octicon-link"></span></a>Uses</h2>

<ul class="list-a">
  <li>Assess and compare the quality of assemblies quickly and visually.</li>
  <li>Identify problematic regions of an assembly.</li>
  <li>Resolve ambiguities in the graph to improve or complete <i>de novo</i> assemblies.</li>
  <li>Determine which other nodes have sequences that are contiguous with a node of interest.</li>
  <li>Annotate graph images to illustrate assembly features.</li>
  <li>Extract candidate sequences that extend through multiple nodes.</li>
</ul>


<h2>
<a id="installation" class="anchor" href="#installation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Installation</h2>

<p>Bandage works equally well on Windows, OS X and Linux. Users are encouraged to download 64-bit binary executables using the links on this page or from the <a href="https://github.com/rrwick/Bandage/releases/" target="_blank">GitHub 'Releases' page</a>. No installation is necessary â€“ just unzip and run. Windows and Mac binaries come packaged with all necessary libraries. The Linux binary has a couple of dependencies, specified in the zip file.</p>

<p>When first opening Bandage on a Mac, you may receive a warning stating that Bandage 'can't be opened because it is from an unidentified developer.'  Right click on the file and select 'Open' to override this warning.</p>

<p>Alternatively, you can clone the source code from GitHub and build Bandage yourself. Instructions are on <a href="https://github.com/rrwick/Bandage/" target="_blank"> the GitHub page</a>.</p>


<h2>
<a id="sampledata" class="anchor" href="#sampledata" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sample data</h2>

<p>Here are a couple of sample graph files you can download and view in Bandage:</p>

<ul class="list-a">
<li><a href="samples/E_coli_LastGraph.zip" target="_blank">An <i>E. coli</i> assembly from Velvet</a>. The reads and more information are available <a href="http://www.ncbi.nlm.nih.gov/sra?linkname=bioproject_sra_all&from_uid=272917" target="_blank">here</a>.</li>
<li><a href="samples/Trinity.fasta.zip" target="_blank">A small mouse transcriptome assembly from Trinity</a>. This data comes with Trinity in the 'sample_data/test_Trinity_Assembly/' folder.</li>
</ul>

<h2>
<a id="example" class="anchor" href="#example" aria-hidden="true"><span class="octicon octicon-link"></span></a>Example</h2>

<p>For a simple case, imagine a bacterial genome that contains a single repeated element in two separate places in the chromosome. A researcher (who does not yet know the structure of the genome) sequences it, and the resulting 100 bp reads are assembled with a <i>de novo</i> assembler.</p>


<div style="text-align: center;">
<div class="diagram" style="max-width:100%; margin-bottom: 20px;">
<img src="images/simple1.png" class="diagram"><br>
</div>
</div>

<p>Because the repeated element is longer than the sequencing reads, the assembler was not able to reproduce the original genome as a single contig. Rather, three contigs are produced: one for the repeated sequence (even though it occurs twice) and one for each sequence between the repeated elements.</p>

<p>Given only the contigs, the relationship between these sequences is not clear. However, the assembly graph contains additional information which is made apparent in Bandage. There are two possible underlying sequences compatible with this graph: two separate circular sequences that share a region in common, or a single larger circular sequence with an element that occurs twice.</p>

<div style="text-align: center;">

<div class="diagram" style="max-width:25%; margin-right: 5%; margin-left: %; margin-bottom: 20px;">
<img src="images/simple2.png" class="diagram"><br>
<small-green>The assembly graph, as it appears in Bandage</small-green>
</div>

<div class="diagram" style="max-width:55%; margin-right: 5%; margin-left: 5%; margin-bottom: 20px;">
<img src="images/simple3.png" class="diagram"><br>
<small-green>Possible underlying sequences compatible with the assembly graph</small-green>
</div>

</div>


<p>Additional knowledge, such as information on the approximate size of the bacterial chromosome, can help the researcher to rule out the first alternative. In this way, Bandage has assisted in turning a fragmented assembly of three contigs into a completed genome of one sequence.</p>



<h2>
<a id="videos" class="anchor" href="#videos" aria-hidden="true"><span class="octicon octicon-link"></span></a>Videos</h2>

<div class="wrapper" style="margin-bottom: 20px;">
    <div class="h_iframe">
        <img class="ratio" src="images/16x9.gif"/>
        <iframe src="https://www.youtube.com/embed/cierloa5hS0?rel=0" frameborder="0" allowfullscreen></iframe>
    </div>
</div>

<div class="wrapper" style="margin-bottom: 20px;">
    <div class="h_iframe">
        <img class="ratio" src="images/16x9.gif"/>
        <iframe src="https://www.youtube.com/embed/rzdQq663lQw?rel=0" frameborder="0" allowfullscreen></iframe>
    </div>
</div>

<div class="wrapper" style="margin-bottom: 20px;">
    <div class="h_iframe">
        <img class="ratio" src="images/16x9.gif"/>
        <iframe src="https://www.youtube.com/embed/VuRN28XyFcI?rel=0" frameborder="0" allowfullscreen></iframe>
    </div>
</div>


<h2>
<a id="screenshots" class="anchor" href="#screenshots" aria-hidden="true"><span class="octicon octicon-link"></span></a>Screenshots</h2>

<p>These graphs are the same bacterial isolate assembled in SPAdes using three different k-mer values:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot01.png" target="_blank">
<img src="images/screenshots/screenshot01_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>K-mer of 21. This value is too small, resulting in short contigs and many connections, giving a dense tangled graph.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot02.png" target="_blank">
<img src="images/screenshots/screenshot02_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>K-mer of 77. This is a good balance, giving long contigs that are well connected.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot03.png" target="_blank">
<img src="images/screenshots/screenshot03_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>K-mer of 127. This value is too large, resulting in the graph breaking into many discontinuous pieces.</small-green>
</div>

<p></p>
<p>The following two images illustrate the difference between single and double nodes:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot04.png" target="_blank">
<img src="images/screenshots/screenshot04_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Single nodes, each holding both forward and reverse sequences.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot05.png" target="_blank">
<img src="images/screenshots/screenshot05_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Double nodes, where forward and reverse nodes are drawn separately.</small-green>
</div>

<p></p>
<p>Examples of custom colouring and labelling of nodes in Bandage:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot06.png" target="_blank">
<img src="images/screenshots/screenshot06_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>A bacterial transposon that shares a gene with the cloning vector.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot07.png" target="_blank">
<img src="images/screenshots/screenshot07_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Assembly graph from metagenomic reads that align to particular 16S genes.</small-green>
</div>

<p></p>
<p>A very large graph from a metagenome assembly, displayed in Bandage:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot08.png" target="_blank">
<img src="images/screenshots/screenshot08_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Zoomed out view, showing large-scale structure.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot09.png" target="_blank">
<img src="images/screenshots/screenshot09_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Zoomed in view of smaller subgraphs.</small-green>
</div>

<p></p>
<p>A BLAST search for a 58 kb transposon in a bacterial assembly:</p>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot10.png" target="_blank">
<img src="images/screenshots/screenshot10_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>The entire assembly graph.</small-green>
</div>

<div class="screenshot_thumbnail" style="width: 210px;">
<a href="images/screenshots/screenshot11.png" target="_blank">
<img src="images/screenshots/screenshot11_thumb.png" class="screenshot_thumbnail"></a><br>
<small-green>Reduced scope to only nodes containing BLAST hits.</small-green>
</div>


<h2>
<a id="cite" class="anchor" href="#cite" aria-hidden="true"><span class="octicon octicon-link"></span></a>Cite</h2>

<p>If you use Bandage in your research, please cite the following publication:</p>
<p><a href="http://bioinformatics.oxfordjournals.org/content/early/2015/07/11/bioinformatics.btv383">Wick R.R., Schultz M.B., Zobel J. & Holt K.E. (2015), Bandage: interactive visualisation of <em>de novo</em> genome assemblies, <em>Bioinformatics</em>.</a> </p>


        </section>

        <aside id="sidebar">
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.6.3/Bandage_Linux_dynamic_v0.6.3.zip" class="button_linux">
            <small>Download</small>
            Linux
          </a>
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.6.3/Bandage_Mac_v0.6.3.zip" class="button_mac">
            <small>Download</small>
            Mac
          </a>
          <a href="https://github.com/rrwick/Bandage/releases/download/v0.6.3/Bandage_Windows_v0.6.3.zip" class="button_windows">
            <small>Download</small>
            Windows
          </a>
		  <a href="https://github.com/rrwick/Bandage" class="button_github">
		    <small>View project on</small>
			GitHub
	      </a>

          <p class="repo-owner"><a href="https://github.com/rrwick/Bandage"></a>Bandage is made by <a href="https://github.com/rrwick">Ryan Wick</a>.</p>
        </aside>
      </div>
    </div>

            <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-59453959-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
